system/Model.php

```diff
@@ -1473,62 +1469,6 @@ protected function cleanValidationRules(array $rules, array $data = null): array
 		return $rules;
 	}
 
-	/**
-	 * Replace any placeholders within the rules with the values that
-	 * match the 'key' of any properties being set. For example, if
-	 * we had the following $data array:
-	 *
-	 * [ 'id' => 13 ]
-	 *
-	 * and the following rule:
-	 *
-	 *  'required|is_unique[users,email,id,{id}]'
-	 *
-	 * The value of {id} would be replaced with the actual id in the form data:
-	 *
-	 *  'required|is_unique[users,email,id,13]'
-	 *
-	 * @param array $rules
-	 * @param array $data
-	 *
-	 * @return array
-	 */
-	protected function fillPlaceholders(array $rules, array $data): array
```

Comment:
I think direct remove protected method is BC break. To keep it, it need to be kept with with change like the following:

```php
    /*
     * @deprecated use fillPlaceholders($data) from Validation instead
     */
    protected function fillPlaceholders(array $rules, array $data): array
    {
        return (function ($validation, $data) {
            $validation->fillPlaceholders($data);
            return $validation->getRules();
        })->bindTo($validation = $this->validation, get_class($this->validation))($validation, $data);
    }
```

Reply:
No other functions in CodeIgniter call fillPlaceholders. You wan't this in case someone extended Model?

Reply:
That's the cost of `protected` method instead of `private`. Ok, let's the maintainer decide to keep or remove it ;)

Reply:
Yeah, I have to agree with @samsonasik here. Anyone extending could use it directly. Even though not too many probably are, we should leave it and add the deprecated note, I think.